### ISSUE TITLE
[iOS only] Can't play HLS faster than 1.0x

### ISSUE THREAD

**Issue by xigxag-dev** [ryanheise/just_audio#210]
**Which API doesn't behave as documented, and how does it misbehave?**
Name here the specific methods or fields that are not behaving as documented, and explain clearly what is happening.
`setSpeed(double speed)`
On iOS only, when playing an HLS stream the playback is capped to 1.0x forward (but < 1.0x works).
I suspect it's related to this post: `https://developer.apple.com/library/archive/qa/qa1772/_index.html` which seems to suggest that the test `_player.currentItem.canPlayFastForward` in the `setSpeed`  method of `AudioPlayer.m` will always return NO. This test is for speeds above 2.0x (I am only googling here, I have no iOS experience!).

**Minimal reproduction project**
The example, but using an HLS source rather than an mp3 source

**To Reproduce (i.e. user steps, not code)**

Replace one of the sample URL's with an HLS stream, try and adjust play speed. Speed will only slow, not increase

**Error messages**

None

**Expected behavior**
Speed up to 2.0x should be possible
Function works correctly on Android

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [e.g. MacOS + version]
 - Browser [e.g. chrome, safari + version]

**Smartphone (please complete the following information):**
 - Device: iPhone X
 - OS: iOS 13.6

**Flutter SDK version**
```
[✓] Flutter (Channel stable, 1.22.1, on Mac OS X 10.15.6 19G2021, locale en-GB)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
[✓] Android Studio (version 4.0)
[✓] Connected device (3 available)

```

**Additional context**

**Comment by ryanheise** [comment]
Yes, it is my understanding from the Apple documentation that you can't exceed speeds of 1.0 for certain types of streams, so your expected behaviour is not actually possible. I have just updated the documentation for this method as follows:

```dart
  /// Sets the playback speed of this player, where 1.0 is normal speed. Note
  /// that on iOS, calls to [setSpeed] may ignored if the underlying media
  /// doesn't support speed changes (e.g. live streams). On Android, values in
  /// excess of 1.0 may result in stalls if the playback speed is faster than
  /// the player is able to downloaded the audio (this should not happen with
  /// locally-stored audio files).
```

**Comment by xigxag-dev** [comment]
Hi - in my application the stream is not a live stream (I can understand that you can't fast forward a live stream).  If I remove the `_player.currentItem.canPlayFastForward` check then my test stream plays fine on iOS at up to 2.0x.  I want parity of features between iOS and Android so I think I will have to fork this and remove that check for now.

**Comment by ryanheise** [comment]
I see, I'll take another look in the morning.

**Comment by ryanheise** [comment]
It will help if you can provide a reproduction project (with a specific HLS URL that I should try, since not all HLS media is the same.)

**Comment by xigxag-dev** [comment]
Hi - It's tough to find audio only, non-live HLS streams on the web and my application server needs authentication so I made a local stream to test with the sample app.
Here's how I tested: If you pull the HLS stream components from here [https://github.com/xigxag-dev/hlstest.git](https://github.com/xigxag-dev/hlstest.git). Then fire up a local http server in that folder - I used `python -m http.server 8080` and then add to the project this AudioSource.  Obviously replace _192.18.1.151_ with the device you are running the server on.
```
HlsAudioSource(
      Uri.parse("http://192.168.1.151:8080/prog_index.m3u8"),
      tag: AudioMetadata(
        album: "hls test",
        title: "hls test",
        artwork:
        "https://media.wnyc.org/i/1400/1400/l/80/1/ScienceFriday_WNYCStudios_1400.jpg",
      ),
    ),
```
You also have to enable `NSAllowsArbitraryLoads` in info.plist 
I think you have said before that you don't have a physical iPhone, so it's possible that you can't use this approach? Sorry that it's a complicated but it was the fastest way for me to test and prove out the function.
All the best.

**Comment by ryanheise** [comment]
Can you tell me any streams on https://bitmovin.com/mpeg-dash-hls-examples-sample-streams/ match your use case?

**Comment by xigxag-dev** [comment]
Hi - unfortunately only 2 of those test streams are live and both are video streams. 
I have created a bucket on AWS with my test stream data, it works fine as a "server" (and is https so you don't need the `NSAllowsArbitraryLoads` entry) - the URI is here [https://xigxag-hls-test-bucket.s3-eu-west-1.amazonaws.com/prog_index.m3u8](https://xigxag-hls-test-bucket.s3-eu-west-1.amazonaws.com/prog_index.m3u8).  

This is pretty representative of my use-case streams (except I use encrypted streams) and you should be able to test with this OK and see that it works fine on Android (or a desktop VLC player) and fine on iOS if you remove the `canPlayFastForward `checker.

Thanks for helping out.

**Comment by ryanheise** [comment]
Thanks, I'll check it out in the morning. I'm not sure yet what would be the best way to handle this at the API level. If I remove the `canPlayFastForward` check from the code, I could expose this API as another method (an iOS-only method). On the other hand, Android's behaviour is that you just try to set the speed and then observe to see whether the speed actually changed (rather than query in advance whether it can be changed).

I'd like to do some experimentation to see if iOS can work in the same way as that, and then I could provide a consistent API for iOS and Android.

**Comment by xigxag-dev** [comment]
I find the Apple docs quite confusing and hard to follow but this link [https://developer.apple.com/library/archive/qa/qa1772/_index.html](https://developer.apple.com/library/archive/qa/qa1772/_index.html) seems to be official Apple documentation and suggests that `canPlayFastForward` is not always a test of can play at up to 2.0x.

**Comment by ryanheise** [comment]
Wow, that is a surprising fact to find hidden away in a Q&A rather than in the API docs.

In short, `canPlayFastForward` isn't talking about > 1.0x it's talking about > 2.0x. I wonder if there are any similar nuances for `canPlaySlowForward`?

I was aware that these queries don't work until the item is ready to play although checking the code I see some extra care also needs to be taken there. There is also the fact that different items in the playlist may have different playback rate capabilities and so when transitioning to an item that can't support the current speed, the plugin should probably automatically clip the speed and broadcast the speed change. This in turn requires me to include the speed in the playback events.

**Comment by xigxag-dev** [comment]
So you also think that's official Apple info? That's certainly how I read it. The developer page for `canPlayFastForward` references "greater than 1.0x" so there is a contradiction here.  Selfishly, for my use case I know all playlist items are derived from the same source so if I "qualify 2.0x playback" then I shouldn't need to test this again. 
I also note there is a `isPlaybackLikelyToKeepUp` bool.  But looking at the docs it also throws a false negative and you have to "make a call on" whether to continue or not.

**Comment by ryanheise** [comment]
Sincere apologies for the delay. I just tested out the URL you provided and found that it supported speeds of even greater than 2.0x, so `canPlayFastForward` seems completely unreliable. It would then probably not be worth exposing an API for `canPlayFastForward` and `canPlaySlowForward` because these properties are, well, seemingly useless.

While not ideal, it does make the fix relatively simple. I've committed to git a change that will allow `setSpeed` to be called regardless of the values given by these two properties.

**Comment by xigxag-dev** [comment]
That's great and please don't apologise for a delay, thanks for helping out with this so quickly!

**Comment by ryanheise** [comment]
The fix has now been published in release 0.5.5.

**Comment by ryanheise** [comment]
I have stumbled upon `AVPlayerItem.timebase` which gives the actual speed of playback which is distinguishable from the requested speed of playback.

https://developer.apple.com/documentation/avfoundation/avplayeritem/1387605-timebase?language=objc
https://developer.apple.com/documentation/coremedia/cmtimebase?language=objc

Something to revisit later.

**Comment by ryanheise** [comment]
Notes: The difference between the requested and actual rate can be due to the capability of the media but it could also be temporary due to buffering/stalling, so care must be taken not to incorrectly report the capability.
